op-node/rollup/interop/managed/system.go

```diff
@@ -378,60 +378,121 @@ func (m *ManagedMode) Reset(ctx context.Context, lUnsafe, xUnsafe, lSafe, xSafe,
+		// in the walkback loop section below if necessary.
+
+		// Search space:
+		// ------------------------------------------------------------------------------------
+		// target.Number |  offset=0   offset=1   offset=2  ...  offset = x-1 = latestUnsafe   |  offset=x  (doesn't exist)
+		// false         |  t/f        t/f        t/f       ...  t/f                           |  true
+		// ------------------------------------------------------------------------------------
+		offset, err := localUnsafeSearch(x, func(i int) (bool, eth.L2BlockRef, error) {
+			return m.isInvalidBlock(ctx, logger, target+1+uint64(i))
+		})
+		if err != nil {
+			return eth.L2BlockRef{}, err
+		}
+		// offset == 0 => search returns 0 if f(i) returned true for all i in [0, x) => all blocks we checked are invalid
+		// offset == x => search returns x if f(i) returned false for all i in [0, x) => all blocks we checked are valid
+
+		if offset != 0 {
+			logger.Info("Found last valid block with binary search", "validNum", target+uint64(offset))
+
+			valid, err := m.l2.L2BlockRefByNumber(ctx, target+uint64(offset))
+			if err != nil {
+				return eth.L2BlockRef{}, err
+			}
+
+			logger.Info("Last valid L2 block ref", "valid", valid)
+			return valid, nil
+		}
+	} else if x < 0 {
+		logger.Warn("Latest unsafe block is older than target, using latest unsafe for search")
+		target = latestUnsafe.Number
 	}
 
-	// Start from latestUnsafe and go down to find the latest valid block, since L1 reorgs aren't usually that deep.
-	for n := latestUnsafe.Number; ; n-- {
-		if n == l2UnsafeTarget.Number-1 {
+	// In the following walkback loop, the following two cases are covered:
+	// 1. x == 0 or x < 1 (i.e. target == latestUnsafe), or
+	// 2. binary search returned offset == 0 (i.e. all blocks we checked are invalid), so we have to go from `target` backwards indefinitely
+	//    until we find a valid block
+	for n := target; ; n-- {
+		if n == target-1 {
 			logger.Warn("No valid unsafe block found up to target, searching further")
 		}
-		// Exit loop if context is done - other errors are only logged and loop continues.
-		if cerr := ctx.Err(); cerr != nil {
-			return eth.L2BlockRef{}, fmt.Errorf("context done: %w", cerr)
-		}
 
-		current, err := m.l2.L2BlockRefByNumber(ctx, n)
+		isInvalid, valid, err := m.isInvalidBlock(ctx, logger, n)
 		if err != nil {
-			logger.Warn("Failed to get L2 block ref, trying previous", "err", err, "current", n)
-			continue
+			return eth.L2BlockRef{}, err
 		}
 
-		logger := logger.New("current", current, "currentL1Origin", current.L1Origin)
-		// make sure L1 origin hasn't been reorged
-		l1Origin, err := m.l1.L1BlockRefByNumber(ctx, current.L1Origin.Number)
+		if !isInvalid {
+			logger.Info("Fould last valid block", "valid", valid)
+			return valid, nil
+		}
+	}
+}
+
+// localUnsafeSearch is a copy-paste from sort.Search, but also stops on and returns error
+func localUnsafeSearch(n int, f func(int) (bool, eth.L2BlockRef, error)) (int, error) {
```

Comment:
The function is a generic binary search, so maybe do
```suggestion
func binarySearch[T any](n int, f func(int) (bool, T, error)) (int, T, error) {
```
and just modify the returns to return the `T` that `f` returned, inside the loop, and just a `var T` at the end. Does that work? This way, you pass through the found element and don't need to fetch it again, but the algorithm stays the same.

Reply:
This particular version of binary search (same as `sort.Search`) is returning the first element for a given f(i), where f(i) is true.

In our context f(i) is `isInvalidBlock`, assuming our search space is something like:

`false, false, false, true, true true`

Therefore I can make it generic, but I cannot use the return value, as we are interested in the element on the left-hand side of the border, not the one returned by `binarySearch[T any]`

-----

We either introduce our own binary search, or we fetch the correct index (I am doing the second).

Reply:
I see the one-off problem due to the binary search direction we're using. So if we used it on an inverted index range, we could pass it `isValidBlock` (instead of _invalid_) and would get the correct element, right?
But I'm ok with the extra fetching now, since we don't want to keep this for long anyways.

Reply:
Copying from Slack, I mean something like
```
invOffset, err := localUnsafeSearch(targetDiff, func(i int) (bool, eth.L2BlockRef, error) {
			block, err := m.verifyBlock(ctx, logger, localUnsafe.Number - uint64(i))
			return block != (eth.L2BlockRef{}), block, err
		})
```